crates/stack/src/network.rs

```diff
@@ -252,53 +314,504 @@ impl Network {
+        Network::new(ip.to_string(), Stack::with(iface))
+    }
+
+    fn produce_data<S, E>(
+        mut tx: S,
+        total: usize,
+        chunk_size: usize,
+    ) -> oneshot::Receiver<anyhow::Result<Vec<u8>>>
+    where
+        S: Sink<Vec<u8>, Error = E> + Unpin + 'static,
+        E: Into<anyhow::Error>,
+    {
+        let (dtx, drx) = oneshot::channel();
+
+        spawn_local(async move {
+            let mut digest = sha3::Sha3_224::new();
+            let mut sent = 0;
+            let mut err = None;
+
+            while sent < total {
+                let vec: Vec<u8> = (0..chunk_size.min(total - sent))
+                    .map(|_| rand::random())
+                    .collect();
+
+                digest.input(&vec);
+                sent += vec.len();
+
+                if let Err(e) = tx.send(vec).await {
+                    err = Some(e);
+                    break;
+                }
+            }
+
+            let _ = tx.close().await;
+
+            println!("produced {} B", sent);
+            match err {
+                Some(e) => dtx.send(Err(e.into())),
+                None => dtx.send(Ok(digest.result().to_vec())),
+            }
+        });
+
+        drx
+    }
+
+    fn consume_data<S, A>(mut rx: S, total: usize) -> oneshot::Receiver<anyhow::Result<Vec<u8>>>
+    where
+        S: Stream<Item = A> + Unpin + 'static,
+        A: AsRef<[u8]>,
+    {
+        let (dtx, drx) = oneshot::channel();
+
+        spawn_local(async move {
+            let mut digest = sha3::Sha3_224::new();
+            let mut read: usize = 0;
+
+            while let Some(vec) = rx.next().await {
+                let len = vec.as_ref().len();
+
+                read += len;
+                digest.input(&vec);
+
+                if read >= total {
+                    break;
+                }
+            }
+
+            println!("consumed {} B", read);
+            let _ = dtx.send(Ok(digest.result().to_vec()));
+        });
+
+        drx
+    }
+
+    fn net_inject<S>(rx: S, net: Network)
+    where
+        S: Stream<Item = Vec<u8>> + 'static,
+    {
+        spawn_local(async move {
+            rx.for_each(|vec| {
+                let net = net.clone();
+                async move {
+                    net.receive(vec);
+                    net.poll();
+                }
+            })
+            .await;
+        });
+    }
+
+    fn net_send<S>(rx: S, net: Network, conn: Connection)
+    where
+        S: Stream<Item = Vec<u8>> + 'static,
+    {
+        spawn_local(async move {
+            let net = net.clone();
+            rx.for_each(|vec| async {
+                match net.send(vec, conn) {
+                    Ok(fut) => {
+                        if let Err(e) = fut.await {
+                            eprintln!("failed to send packet: {}", e);
+                        }
+                    }
+                    Err(e) => eprintln!("failed to send packet: {}", e),
+                }
```

Comment:
`Queue` was introduced here with a purpose - synchronous ordering of data to be sent asynchronously. Making `net.send` async would be syntactic sugar on top of awaiting the otherwise returned future. 

Spawning multiple async `net.send`s would still not guarantee ordering since the inner `Send` futures (there may be multiple e.g. due to payload chunking) may interleave. Also, event loop submission order does not guarantee the order in which the futures are processed,